### PR TITLE
feat: update metadata extraction code

### PR DESCRIPTION
**Pull request by ebezzi**
### Reviewers
**Functional:** 
@MDunitz @seve 

**Readability:** 
@leslieypark 
---

Zenhub Tickets: 
[#1314](https://app.zenhub.com/workspaces/single-cell-5e2a191dad828d52cc78b028/issues/chanzuckerberg/corpora-data-portal/1314)
[#1300](https://app.zenhub.com/workspaces/single-cell-5e2a191dad828d52cc78b028/issues/chanzuckerberg/corpora-data-portal/1300)
[#1315](https://app.zenhub.com/workspaces/single-cell-5e2a191dad828d52cc78b028/issues/chanzuckerberg/corpora-data-portal/1315)
[#1299](https://app.zenhub.com/workspaces/single-cell-5e2a191dad828d52cc78b028/issues/chanzuckerberg/corpora-data-portal/1299)

## Changes
- #1314: Adds the new fields to `extract_metadata`, which is used to read an AnnData file and insert the metadata for the dataset in the Postgres database.
- #1300: Modify `extract_metadata` logic to replace `layer_descriptions` with new schema 2.0.0 `X_normalization`
- #1315: Modify `get_corpora_properties()`, `REQUIRED_SIMPLE_METADATA_FIELDS`, and related code to be schema 2.0.0 compliant.
- #1299: Modify `get_corpora_versions()` and related code to be schema 2.0.0 compliant.

## Definition of Done (from ticket)
DOD not in the ticket, but the changes are limited in scope

## QA steps (optional)
None yet since this is part of a feature branch.

## Known Issues
N/A

### CODE REVIEW

**Comment by tihuan** [comment]
please add ticket number in the PR description e.g., `#1234` or click connect issue in Zenhub thanks!

File: backend/corpora/dataset_processing/process.py [hunk 1]
```diff
@@ -284,15 +283,19 @@ def _get_term_pairs(base_term):
...
         "tissue": _get_term_pairs("tissue"),
         "assay": _get_term_pairs("assay"),
         "disease": _get_term_pairs("disease"),
-        "sex": list(adata.obs.sex.unique()),
+        "sex": _get_term_pairs("sex"),
         "ethnicity": _get_term_pairs("ethnicity"),
         "development_stage": _get_term_pairs("development_stage"),
         "cell_count": adata.shape[0],
         "mean_genes_per_cell": numerator / denominator,
+        "is_primary_data": adata.obs["is_primary_data"].unique()[0],
```
Comment (ebezzi) [inline backend/corpora/dataset_processing/process.py:295]: TODO: https://app.zenhub.com/workspaces/single-cell-5e2a191dad828d52cc78b028/issues/chanzuckerberg/corpora-data-portal/1356. Will be done in a separate PR

File: backend/corpora/common/utils/corpora_constants.py [hunk 2]
```diff
@@ -80,11 +80,9 @@ class CorporaConstants(object):
     ]
 
     REQUIRED_SIMPLE_METADATA_FIELDS = [
-        "version",
```
Comment (seve) [inline backend/corpora/common/utils/corpora_constants.py:83]: do we want to replace this with `schema_version`?

File: backend/corpora/dataset_processing/h5ad_data_file.py [hunk 3]
```diff
@@ -289,34 +287,22 @@ def get_corpora_properties(self):
...
-
-    def corpora_is_version_supported(self, corpora_schema_version, corpora_encoding_version):
-        return (
-            corpora_schema_version
-            and corpora_encoding_version
-            and corpora_schema_version.startswith("1.")
-            and corpora_encoding_version.startswith("0.1.")
-        )
+    def corpora_is_schema_version_supported(self, schema_version):
+        return schema_version and (schema_version.startswith("1.") or schema_version.startswith("2."))
```
Comment (seve) [inline backend/corpora/dataset_processing/h5ad_data_file.py:308]: I think we can safely assume that every dataset we ingest following the merge of the epic will be schema 2.0.0

Comment (seve) [reply backend/corpora/dataset_processing/h5ad_data_file.py:308]: If not, we should write a ticket and put a TODO here for when we no longer allow 1.x.x datasets

File: backend/corpora/common/utils/corpora_constants.py (hunk 2 above)
Comment (leslieypark) [reply backend/corpora/common/utils/corpora_constants.py:83]: Missed that - thank you!

File: backend/corpora/dataset_processing/process.py [hunk 4]
```diff
@@ -254,24 +254,22 @@ def extract_metadata(filename):
...
-    try:
-        raw_layer_name = [k for k, v in adata.uns["layer_descriptions"].items() if v == "raw"][0]
-    except (KeyError, IndexError):
-        raise RuntimeError("Raw layer not found in layer descriptions!")
-
-    if raw_layer_name == "X":
-        raw_layer = adata.X
-    elif raw_layer_name == "raw.X":
-        raw_layer = adata.raw.X
+    # TODO: Concern with respect to previous use of raising error when there is no raw layer.
```
Comment (MDunitz) [inline backend/corpora/dataset_processing/process.py:257]: I'm not clear on what task this todo is referencing

File: tests/unit/backend/corpora/dataset_processing/test_process.py [hunk 5]
```diff
@@ -104,10 +104,8 @@ def test_extract_metadata(self, mock_read_h5ad):
             numpy.random.randint(10, size=(50001, 5)) * 50, columns=list("ABCDE"), index=(str(i) for i in range(50001))
         )
 
-        tissue = numpy.random.choice([0, 1], size=(50001))
-        assay = numpy.random.choice([0, 1, 2], size=(50001))
-        eth = numpy.random.choice([0, 1], size=(50001))
-        dev = numpy.random.choice([0, 1, 2], size=(50001))
+        eth = tissue = numpy.random.choice([0, 1], size=(50001))
```
Comment (MDunitz) [inline tests/unit/backend/corpora/dataset_processing/test_process.py:107]: what is eth?

File: tests/unit/backend/corpora/dataset_processing/test_process.py [hunk 6]
```diff
@@ -104,10 +104,8 @@ def test_extract_metadata(self, mock_read_h5ad):
...
+        assay = dev = sx = numpy.random.choice([0, 1, 2], size=(50001))
```
Comment (MDunitz) [inline tests/unit/backend/corpora/dataset_processing/test_process.py:108]: can you use a more informative variable name? I'm assuming sx is for sex?

File: tests/unit/backend/corpora/dataset_processing/test_process.py [hunk 7]
```diff
@@ -239,22 +247,31 @@ def test_extract_metadata_find_raw_layer(self, mock_read_h5ad):
...
             ],
             index=(str(i) for i in range(11)),
         )
         uns = {
             "title": "my test dataset",
-            "organism": "Homo sapiens",
-            "organism_ontology_term_id": "NCBITaxon:8505",
-            "layer_descriptions": {"my_awesome_wonky_layer": "raw"},
+            "X_normalization": "normal",
+            "X_approximate_distribution": "normal",
```
Comment (MDunitz) [inline tests/unit/backend/corpora/dataset_processing/test_process.py:266]: are we testing the logic for "count" as well?

File: backend/corpora/dataset_processing/process.py (hunk 4 above)
Comment (ebezzi) [reply backend/corpora/dataset_processing/process.py:257]: See the discussion [here](https://github.com/chanzuckerberg/single-cell-data-portal/pull/1358).

File: tests/unit/backend/corpora/dataset_processing/test_process.py (hunk 7 above)
Comment (ebezzi) [reply tests/unit/backend/corpora/dataset_processing/test_process.py:266]: No, but there is no real logic behind `X_approximate_distribution`, it's just a constant value that is passed through as is.

File: tests/unit/backend/corpora/dataset_processing/test_process.py (hunk 5 above)
Comment (ebezzi) [reply tests/unit/backend/corpora/dataset_processing/test_process.py:107]: I believe it's for `ethnicity`

File: backend/corpora/dataset_processing/process.py (hunk 4 above)
Comment (MDunitz) [reply backend/corpora/dataset_processing/process.py:257]: Sorry I'm still a little confused. Is this a code change you are recommending? or just making a note of a particular issue?

Comment (ebezzi) [reply backend/corpora/dataset_processing/process.py:257]: Previously the code used to raise an error if a raw layer wasn't present. The new code defaults to `AnnData.X` if the raw data isn't available. This is a significant change and was marked as a concern. That said, the product team seems onboard with the change, but I would like @pablo-gar to also confirm he's fine with it before we commit it. That said, I think adjustments can be made as part of a separate PR.

File: tests/unit/backend/corpora/dataset_processing/test_process.py (hunk 5 above)
Comment (ebezzi) [reply tests/unit/backend/corpora/dataset_processing/test_process.py]: Renamed

File: tests/unit/backend/corpora/dataset_processing/test_process.py (hunk 6 above)
Comment (ebezzi) [reply tests/unit/backend/corpora/dataset_processing/test_process.py]: Done

File: tests/unit/backend/corpora/dataset_processing/test_process.py (hunk 7 above)
Comment (ebezzi) [reply tests/unit/backend/corpora/dataset_processing/test_process.py:270]: Can you point to where this is used? I can try to add some unit tests there.

File: tests/unit/backend/corpora/dataset_processing/test_process.py (hunk 5 above)
Comment (MDunitz) [reply tests/unit/backend/corpora/dataset_processing/test_process.py]: Thanks!

File: tests/unit/backend/corpora/dataset_processing/test_process.py (hunk 6 above)
Comment (MDunitz) [reply tests/unit/backend/corpora/dataset_processing/test_process.py]: Thanks!

File: tests/unit/backend/corpora/dataset_processing/test_process.py (hunk 7 above)
Comment (MDunitz) [reply tests/unit/backend/corpora/dataset_processing/test_process.py:278]: I think there are should be unit tests for the two possible enums https://github.com/chanzuckerberg/single-cell-explorer/blob/main/backend/common/compute/diffexp_generic.py#L33

**Review by MDunitz — APPROVED** [review]
lgtm

**Review by seve — APPROVED** [review]
